packages/sdk/src/modules/Vaults.ts

```diff
@@ -0,0 +1,225 @@
+                totalSupply,
+                totalSupplyNative,
+                asset,
+                symbol,
+                supplyApy,
+                adapterCount,
+                emergencyExit,
+                adapters,
+                decimals,
+                underlyingPrice,
+                extraDocs,
+                performanceFee,
+                depositFee,
+                withdrawalFee,
+                managementFee,
+              };
+            })
+          );
+        } catch (error) {
+          this.logger.error(`get vaults error in chain ${this.chainId}:  ${error}`);
+
+          throw Error(
+            `Getting vaults failed in chain ${this.chainId}: ` + (error instanceof Error ? error.message : error)
+          );
+        }
+      } else {
+        return [];
+      }
+    }
+
+    async vaultApprove(vault: string, asset: string) {
+      const token = getContract(asset, EIP20InterfaceABI, this.signer);
+      const tx = await token.approve(vault, constants.MaxUint256);
+
+      return tx;
+    }
+
+    async convertToShares(amount: BigNumber, vault: string) {
+      const optimizedAPRVault = this.createOptimizedAPRVault(vault);
+
+      const totalSupply = await optimizedAPRVault.callStatic.totalSupply();
+      const totalAssets = await optimizedAPRVault.callStatic.totalAssets();
+
+      return amount.mul(totalSupply.add(utils.parseUnits("1", DECIMAL_OFFSET))).div(totalAssets.add("1"));
+    }
+
+    async vaultDeposit(vault: string, amount: BigNumber) {
+      const optimizedAPRVault = this.createOptimizedAPRVault(vault, this.signer);
+      const response = await optimizedAPRVault.callStatic["deposit(uint256)"](amount);
+      const { deposit: depositFee } = await optimizedAPRVault.callStatic.fees();
+
+      const feeShares = await this.convertToShares(amount.mul(depositFee).div(utils.parseUnits("1")), vault);
+
+      const shares = (await this.convertToShares(amount, vault)).sub(feeShares);
+
+      if (!response.eq(shares)) {
+        const errorCode = parseInt(response.toString());
+
+        return { errorCode };
+      }
+
+      const tx: ContractTransaction = await optimizedAPRVault["deposit(uint256)"](amount);
+
+      return { tx, errorCode: null };
+    }
+
+    async vaultWithdraw(vault: string, amount: BigNumber) {
+      const optimizedAPRVault = this.createOptimizedAPRVault(vault, this.signer);
+      const response = await optimizedAPRVault.callStatic["withdraw(uint256)"](amount);
+      const { withdrawal: withdrawalFee } = await optimizedAPRVault.callStatic.fees();
+
+      const shares = await this.convertToShares(amount, vault);
+      const feeShares = shares.mul(withdrawalFee).div(utils.parseUnits("1").sub(withdrawalFee));
+
+      // if rounded up, then should add "1"
+      if (!(response.eq(shares.add(feeShares)) || response.eq(shares.add("1").add(feeShares)))) {
+        const errorCode = parseInt(response.toString());
+
+        return { errorCode };
```

Comment:
the response is never an error code, so this manual calculation and comparing of the returned value is not good

Reply:
So do I not need to do check with `callStatic` function? and need to directly do `await optimizedAPRVault["deposit(uint256)"](amount);`?

Reply:
confirmed, no need to check with `callStatic`